src/ray/core_worker/transport/direct_task_transport.h

```diff
@@ -199,6 +223,14 @@ class CoreWorkerDirectTaskSubmitter {
 
   // Retries cancelation requests if they were not successful.
   absl::optional<boost::asio::steady_timer> cancel_retry_timer_;
+
+
+  int pipeline_fullness_;
+
+  // Keeps track of how many tasks are in-flight to a given worker, so that we can
+  // make sure to keep the pipeline full
+  absl::flat_hash_map<rpc::WorkerAddress, int> tasks_in_flight_ GUARDED_BY(mu_);
```

Comment:
It'd be nice to combine this into the `worker_to_lease_client_` so that we don't need to have a separate hashmap lookup/cleanup. It probably makes sense to change the value of the existing map to a struct that represents a single executor.